<a href="https://colab.research.google.com/github/kozinofsky/Coursera_Capstone/blob/master/Neural_networks_final_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**README!**
Thank you for reviewing my assignment!
It is consists of **4 parts: A, B, C, D**.
For each part you will see the appropriate title, for example **PART A**. Couple words here, some time I use "mse/MSE", it means Mean Squared Error.

**PART A**



Task description:
In the PART A we need to build a baseline model.
Using the Keras library to build a neural network with the following parameters:
- **One hidden layer** of **10 nodes**, and a **ReLU** activation function
- Use the **adam** optimizer and the **mean squared error **as the loss function.

In [0]:
import keras

Let's upload our data:

In [0]:
import pandas as pd
import numpy as np

In [34]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head() # to show our data set

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

We need to separate our data on predictors and target

In [35]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
target = concrete_data['Strength']

target.head() # now we have two data set, "target" and "predictors", let's see what "target" dataset consists of

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Now we need to split our data like that:
**30% data for testing** and the rest for training.

In [36]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.33)

X_train.head() # now we have 4 dataset and I show for example how "X_train" dataset looks like:

Cement  Blast Furnace Slag  ...  Fine Aggregate  Age
163    313.3               262.2  ...           611.8   91
1000   141.9               166.6  ...           785.3   28
61     266.0               114.0  ...           670.0  270
900    150.0               237.0  ...           675.0   28
652    102.0               153.0  ...           942.0    3

[5 rows x 8 columns]

Now we need to create our regression model, let's create a separate function for that:

In [0]:
import keras 

from keras.models import Sequential
from keras.layers import Dense

In [38]:
data_shape = predictors.shape[1]
data_shape # to find out how much inputs should be in our model

8

The model, as task says, should consist of:
* **one hidden layer** of **10** nodes and **ReLU** activation function
* we should use **adam** optimizer and the **mean squared error** as the loss function

In [0]:
def our_model():
  # create model
  model = Sequential()
  model.add(Dense(10, activation='relu', input_shape = (data_shape,))) # hidden layer with 10 nodes and ReLU as activation function)
  model.add(Dense(1)) # the output layer

  # compile model 
  model.compile(optimizer='adam', loss='mean_squared_error')
  return model

Now we should train our model using 50 epochs

In [0]:
# build the model
model = our_model()
# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)

Now we need to **evaluate our model** on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength.
We will use the **mean_squared_error** function from Scikit-learn.

In [28]:
from sklearn.metrics import mean_squared_error

true_data = target # here is our actual data, from original data set
predicted_data = model.predict(predictors) # getting predicted data
print(predicted_data)
print(predicted_data.shape)
print(true_data)
print(true_data.shape)

mse = mean_squared_error(true_data, predicted_data, squared=True) # mean squared error between actual and predicted data
mse

[[57.783737]
 [58.10766 ]
 [66.4379  ]
 ...
 [28.45022 ]
 [24.545185]
 [35.060516]]
(1030, 1)
0       79.99
1       61.89
2       40.27
3       41.05
4       44.30
        ...  
1025    44.28
1026    31.18
1027    23.70
1028    32.77
1029    32.40
Name: Strength, Length: 1030, dtype: float64
(1030,)


166.2068881449054

Now the task is to repeat all the steps 50 times and get a list of 50 mean squared errors.
Let me remind the steps we did:
1. Randomly split the data into a training and test sets by holding 30% of the data for testing. 
2. Train the model on the training data using 50 epochs
3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's do it!

In [47]:
model = our_model()
true_data = target # actual data
mse_std_list = [] # initialize an empty list for 50 mean squared errors
for i in range(0,50):
  X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=24)
  model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2) # train the model
  
  predicted_data = model.predict(predictors) # getting predicted data, using our trained model
  mse = mean_squared_error(true_data, predicted_data, squared=True) # find out mean squared error for the first iteration
  mse_std_list.append(mse) # add new mean squared error to our list for 50 of them

print(mse_std_list)

Streaming output truncated to the last 5000 lines.
 - 0s - loss: 2724.3505 - val_loss: 2644.2016
Epoch 27/50
 - 0s - loss: 2648.5434 - val_loss: 2578.4519
Epoch 28/50
 - 0s - loss: 2572.6559 - val_loss: 2507.3419
Epoch 29/50
 - 0s - loss: 2504.0118 - val_loss: 2443.0958
Epoch 30/50
 - 0s - loss: 2435.6665 - val_loss: 2375.9847
Epoch 31/50
 - 0s - loss: 2365.5344 - val_loss: 2323.0890
Epoch 32/50
 - 0s - loss: 2303.6209 - val_loss: 2264.9341
Epoch 33/50
 - 0s - loss: 2245.7875 - val_loss: 2206.7375
Epoch 34/50
 - 0s - loss: 2183.0752 - val_loss: 2149.5152
Epoch 35/50
 - 0s - loss: 2125.8406 - val_loss: 2093.8222
Epoch 36/50
 - 0s - loss: 2071.4996 - val_loss: 2046.6026
Epoch 37/50
 - 0s - loss: 2019.5911 - val_loss: 2006.7794
Epoch 38/50
 - 0s - loss: 1966.9085 - val_loss: 1953.1592
Epoch 39/50
 - 0s - loss: 1916.8224 - val_loss: 1910.2657
Epoch 40/50
 - 0s - loss: 1871.2345 - val_loss: 1865.6458
Epoch 41/50
 - 0s - loss: 1825.4557 - val_loss: 1825.7567
Epoch 42/50
 - 0s - loss: 1783.29

So the conclusion here, in the *1 iteration* the **mean squared error** was **162.8**, but *after* our model learnt during 50 iterations, the **mse** value became **37.54** which is much less compare with 162.8 at the first iteration

In [61]:
print(mse_std_list)
print(np.shape(mse_std_list)) # double check how much mean squared errors I have in my list

[1479.8524207027133, 639.2643092836585, 310.86126686290635, 175.11071703307292, 126.9388260259404, 111.06693870754384, 104.92639354456045, 92.40175696186246, 68.43103135957259, 69.92256017874853, 55.30677730693545, 49.30704968054147, 49.85246894307608, 48.680908551070196, 46.58813462172897, 46.4793087794039, 48.17173693860066, 46.40652545288922, 46.440207919314375, 46.091388548305936, 46.30879649940941, 46.206305559738524, 45.61040359118449, 45.70525314811768, 45.49831232033177, 46.89195893473663, 48.03809786923494, 45.68925151726901, 49.03001363573975, 50.19818322864421, 45.40551949036745, 51.0106088349783, 50.160347661868535, 48.67246595454331, 45.625103118845374, 50.97359383039355, 49.20924704338607, 45.65169203818501, 47.1378130851191, 45.185224594841614, 44.817491887232585, 45.48027296238072, 46.08851884853903, 49.710050064224745, 52.98484099747249, 43.73314621110454, 54.063189947783364, 44.41586972793229, 43.43224806829406, 61.45642802776797]
(50,)


In [62]:
print('Standart deviation of mse is ' + str(np.std(mse_std_list)))

Standart deviation of mse is 217.09086089287803


In [63]:
print('The mean of mean squared errors is ' + str(np.mean(mse_std_list)))

The mean of mean squared errors is 101.92981952204224


**Conclusion** for the **PART A**:
We can see, that in the first iteration our MSE was 1479.85. After our model was learning during 50 iterations, the MSE became 61.45.
The **mean** of MSEs (all 50) is 101.9
The **standart deviation** is 217.09.

**PART "B"**

In part "B" we are required to **normalize** our data.
By *subtracting the mean* from the individual predictors and *dividing by the standard deviation*.



In [64]:
predictors_norm = (predictors - predictors.mean()) / predictors.std() # normalizing of predictors
predictors_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

And now we should repeat the whole part "A" but with normalized data.

In [65]:
norm_model = our_model() # creating new model
true_data = target # actual data
mse_list2 = [] # empty list for the list of 50 mean squared errors

for i in range(0,50):
  X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=24)
  # the only different with the PART A is that we use "predictors_norm" set as predictors data
  norm_model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)
  
  predicted_data = norm_model.predict(predictors_norm)
  mse2 = mean_squared_error(true_data, predicted_data, squared=True)
  mse_list2.append(mse2)

print(mse_list2)

print('Standart deviation of MSE is ' + str(np.std(mse_list2)))
print('The mean of mean squared errors is ' + str(np.mean(mse_list2)))

Streaming output truncated to the last 5000 lines.
 - 0s - loss: 1028.8574 - val_loss: 938.0453
Epoch 28/50
 - 0s - loss: 995.0132 - val_loss: 905.9939
Epoch 29/50
 - 0s - loss: 959.9429 - val_loss: 875.4784
Epoch 30/50
 - 0s - loss: 926.7687 - val_loss: 843.9609
Epoch 31/50
 - 0s - loss: 892.9597 - val_loss: 813.2672
Epoch 32/50
 - 0s - loss: 859.7085 - val_loss: 783.4988
Epoch 33/50
 - 0s - loss: 827.3340 - val_loss: 753.7749
Epoch 34/50
 - 0s - loss: 795.2940 - val_loss: 725.4228
Epoch 35/50
 - 0s - loss: 764.6555 - val_loss: 697.4623
Epoch 36/50
 - 0s - loss: 734.5717 - val_loss: 670.7655
Epoch 37/50
 - 0s - loss: 705.4895 - val_loss: 644.6628
Epoch 38/50
 - 0s - loss: 677.5877 - val_loss: 619.6696
Epoch 39/50
 - 0s - loss: 650.8068 - val_loss: 595.6390
Epoch 40/50
 - 0s - loss: 625.1390 - val_loss: 572.5678
Epoch 41/50
 - 0s - loss: 600.3346 - val_loss: 550.6819
Epoch 42/50
 - 0s - loss: 576.8264 - val_loss: 529.3871
Epoch 43/50
 - 0s - loss: 554.0945 - val_loss: 509.2577
Epoch 44

**Conclusion** for **PART B**.
In the *first iteration* the MSE was **410.82**.
After learning during *50 iterations*), *MSE* became **35.00**
The **standart deviation** of mse is **56.44**
and the **mean** of *MSE* is **51**

So, the result for normalize data is much better compare with non normalize data

Let me remind the **results** of **PART A**
* the first iteration, MSE was **1479.85**. 
* After 50 iterations, the MSE became **61.45**.
* The **mean** of MSEs (all 50) is **101.9**
* The **standart deviation** is **217.09**.

**PART C**

In part "C" we are required to increase the number of epochs to 100. And repeat part B after that (with 100 epochs).
All we need to change is number of epochs during our training.

In [67]:
hund_epochs_model = our_model()
true_data = target
mse_list3 = []

for i in range(0,50):
  X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=24)
  hund_epochs_model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=100, verbose=2) #here is our change
  
  predicted_data = hund_epochs_model.predict(predictors_norm)
  mse3 = mean_squared_error(true_data, predicted_data, squared=True)
  #std = np.std()
  mse_list3.append(mse3)
  # and add it to the list

print(mse_list3)

print('Standart deviation of mse when 100 epochs is ' + str(np.std(mse_list3)))
print('The mean of mean squared errors is ' + str(np.mean(mse_list3)))

Streaming output truncated to the last 5000 lines.
Epoch 15/100
 - 0s - loss: 34.5229 - val_loss: 42.7789
Epoch 16/100
 - 0s - loss: 34.4390 - val_loss: 42.5992
Epoch 17/100
 - 0s - loss: 34.4109 - val_loss: 42.6270
Epoch 18/100
 - 0s - loss: 34.3220 - val_loss: 42.7335
Epoch 19/100
 - 0s - loss: 34.3188 - val_loss: 42.7404
Epoch 20/100
 - 0s - loss: 34.3155 - val_loss: 42.4490
Epoch 21/100
 - 0s - loss: 34.3373 - val_loss: 42.5264
Epoch 22/100
 - 0s - loss: 34.5055 - val_loss: 42.8108
Epoch 23/100
 - 0s - loss: 34.3321 - val_loss: 42.6373
Epoch 24/100
 - 0s - loss: 34.4107 - val_loss: 42.4051
Epoch 25/100
 - 0s - loss: 34.3352 - val_loss: 42.6110
Epoch 26/100
 - 0s - loss: 34.3671 - val_loss: 42.8157
Epoch 27/100
 - 0s - loss: 34.3100 - val_loss: 42.6946
Epoch 28/100
 - 0s - loss: 34.3518 - val_loss: 42.4167
Epoch 29/100
 - 0s - loss: 34.4718 - val_loss: 43.0151
Epoch 30/100
 - 0s - loss: 34.2731 - val_loss: 42.3680
Epoch 31/100
 - 0s - loss: 34.3871 - val_loss: 42.6732
Epoch 32/100
 

Conclusion for **PART C**:
* on first iteration the MSE is **160.75**
* after 50 iterations MSE became **35.84**
* Standart deviation is **20.12**
* Mean is **42.80**

Let me remind the results in **PART B**:
* *first iteration*, MSE was **410.82**.
* After *50 iterations*, *MSE* became **35.00**
* **standart deviation** of mse is **56.44**
* **mean** of *MSE* is **51**

The results of PART C is better compared with the results of PART B. So, we can see that increasing of number of epochs positively affects final results.

**PART D**

Here we need to increase the number of hidden layers:
- Three hidden layers, each of 10 nodes and ReLU activation function.

and repeat part B. Small reminder here, in part B we train our model only during 50 epochs, but using the normalized data.

Let's code!

In [68]:
def layers3_model():
  # create model
  model = Sequential()
  model.add(Dense(10, activation='relu', input_shape = (data_shape,))) # firts hidden layer
  model.add(Dense(10, activation='relu')) # second hidden layer
  model.add(Dense(10, activation='relu')) # third hidden layer
  model.add(Dense(1)) # output layer

  # compile model 
  model.compile(optimizer='adam', loss='mean_squared_error')
  return model


more_layers_model = layers3_model()
true_data = target #actual data
mse_list4 = [] # empty list for the list of 50 mean squared errors

for i in range(0,50):
  X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=24)
  more_layers_model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)
  
  predicted_data = more_layers_model.predict(predictors_norm)
  mse4 = mean_squared_error(true_data, predicted_data, squared=True)
  mse_list4.append(mse4) # adding current MSE to the final list of 50 MSEs

print(mse_list4)

print('Standart deviation of MSEs is ' + str(np.std(mse_list4)))
print('The mean of mean squared errors is ' + str(np.mean(mse_list4)))

Streaming output truncated to the last 5000 lines.
 - 0s - loss: 167.9052 - val_loss: 149.0120
Epoch 28/50
 - 0s - loss: 165.6211 - val_loss: 146.2082
Epoch 29/50
 - 0s - loss: 163.1892 - val_loss: 144.3594
Epoch 30/50
 - 0s - loss: 161.5761 - val_loss: 143.0863
Epoch 31/50
 - 0s - loss: 158.7105 - val_loss: 139.7725
Epoch 32/50
 - 0s - loss: 156.5331 - val_loss: 139.3276
Epoch 33/50
 - 0s - loss: 154.2530 - val_loss: 137.5755
Epoch 34/50
 - 0s - loss: 152.0584 - val_loss: 135.2350
Epoch 35/50
 - 0s - loss: 149.6514 - val_loss: 134.9461
Epoch 36/50
 - 0s - loss: 147.2646 - val_loss: 132.1749
Epoch 37/50
 - 0s - loss: 144.8880 - val_loss: 129.1299
Epoch 38/50
 - 0s - loss: 142.4516 - val_loss: 128.4056
Epoch 39/50
 - 0s - loss: 139.2255 - val_loss: 125.4760
Epoch 40/50
 - 0s - loss: 136.2060 - val_loss: 124.0147
Epoch 41/50
 - 0s - loss: 133.0408 - val_loss: 121.4596
Epoch 42/50
 - 0s - loss: 129.8082 - val_loss: 118.7533
Epoch 43/50
 - 0s - loss: 126.2787 - val_loss: 115.8603
Epoch 44/

**Conclusion for the PART D:**
* on the *first iteration* the MSE is **104.89**
* after *50 iterations* MSE became **21.68**
* *Standart devation* of 50 MSEs is **13.27**
* and the *mean* is **28.1**

Let me remind the results of PART B and PART C.

**PART C**:
* on first iteration the MSE is **160.75**
* after 50 iterations MSE became **35.84**
* Standart deviation is **20.12**
* Mean is **42.80**

**PART B**:
* *first iteration*, MSE was **410.82**.
* After *50 iterations*, *MSE* became **35.00**
* **standart deviation** of mse is **56.44**
* **mean** of *MSE* is **51**

We can see, compare with B and C, that 2 additional layers improved the results very strong. 
Even 50 epochs, comparing with 100 in part "B", didn't improve the results so, as two additional hidden layers.